WordPress/Classes/ViewRelated/Post/PostTagPickerViewController.swift

```diff
@@ -0,0 +1,402 @@
+    /// The algorithm here differs slightly as we don't want to interpret the last tag as a partial one.
+    ///
+    fileprivate func normalizeInitialTags(tags: [String]) -> String {
+        var tags = tags.filter({ !$0.isEmpty })
+        tags.append("")
+        return tags.joined(separator: ", ")
+    }
+
+    fileprivate func updateSuggestions() {
+        dataSource.selectedTags = completeTags
+        dataSource.searchQuery = partialTag
+        reloadTableData()
+    }
+}
+
+extension PostTagPickerViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
+        switch tableView.dataSource {
+        case is FailureDataSource:
+            loadTags()
+        case is LoadingDataSource:
+            return
+        case is SuggestionsDataSource:
+            suggestionTapped(cell: tableView.cellForRow(at: indexPath))
+        default:
+            assertionFailure("Unexpected data source")
+        }
+    }
+
+    private func suggestionTapped(cell: UITableViewCell?) {
+        guard let tag = cell?.textLabel?.text else {
+            return
+        }
+        complete(tag: tag)
+    }
+}
+
+// MARK: - Data Sources
+
+private protocol PostTagPickerDataSource: UITableViewDataSource {
+    var selectedTags: [String] { get set }
+    var searchQuery: String { get set }
+}
+
+private class LoadingDataSource: NSObject, PostTagPickerDataSource {
+    var selectedTags = [String]()
+    var searchQuery = ""
+
+    static let cellIdentifier = "Loading"
+    typealias Cell = UITableViewCell
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 1
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: LoadingDataSource.cellIdentifier, for: indexPath)
+        WPStyleGuide.configureTableViewSuggestionCell(cell)
+        cell.textLabel?.text = NSLocalizedString("Loading...", comment: "Loading tags")
+        cell.selectionStyle = .none
+        return cell
+    }
+}
+
+private class FailureDataSource: NSObject, PostTagPickerDataSource {
+    var selectedTags = [String]()
+    var searchQuery = ""
+
+    static let cellIdentifier = "Failure"
+    typealias Cell = UITableViewCell
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 1
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: FailureDataSource.cellIdentifier, for: indexPath)
+        WPStyleGuide.configureTableViewSuggestionCell(cell)
```

Comment:
I kind of missed the message when I tested on Airplane mode though, so I would like to differentiate it a bit from Loading, but yeah, red might be too much...